tests/lib/rules/no-self-assign.js

```diff
@@ -39,7 +39,13 @@ ruleTester.run("no-self-assign", rule, {
         {code: "({a} = {a: b})", parserOptions: {ecmaVersion: 6}},
         {code: "({a} = {a() {}})", parserOptions: {ecmaVersion: 6}},
         {code: "({a} = {[a]: a})", parserOptions: {ecmaVersion: 6}},
-        {code: "({a, ...b} = {a, ...b})", parserOptions: {ecmaVersion: 6, ecmaFeatures: {experimentalObjectRestSpread: true}}}
+        {code: "({a, ...b} = {a, ...b})", parserOptions: {ecmaVersion: 6, ecmaFeatures: {experimentalObjectRestSpread: true}}},
+        "a.b = a.c",
+        "a.b = c.b",
+        "a.b = a[b]",
+        "a[b] = a.b",
+        "a.b().c = a.b().c",
```

Comment:
Nice test case. I'm wondering, should `a().b.c = a().b.c` constitute a self-assignment or not? (`b.c` is at least a constant member expression, but `a()` could return a different object on each invocation, so probably not?)